First pass at implementing archiving functionality

nikilselvam opened the pull request:
This PR looks at implementing 'archive conversation' functionality. The first commit sets up the front-end logic for archiving and links in a POST request to allow the front-end logic to trigger the back-end logic. The second commit is my first pass at filling out the back-end logic.

The back-end logic needs some more work for the functionality to be complete. @psybuzz would you mind taking a look and helping me out with this part in particular? I wrote the parts that I could in `exports.archive` of `routes/conversation.js` using `exports.leave` of the same file as an example, but I wasn't sure how to incorporate in promises as we do for leave functionality (or if that's even necessary here). Moreover, I included psuedo-code for how we would actually modify the hallOfFame boolean for the given conversation but wasn't sure how to actually modify the DB entry properly. Any comments you can provide here would be great!

nikilselvam commented on routes/conversation.js:
```diff
@@ -255,6 +255,39 @@ exports.leave = function (req, res){
+
+		// Modify the hallOfFame boolean associated with the conversation to true if it exists.
+		if (found) {
+			// Something along the lines of: user.userConversations[found].hallOfFame == true;
```
@psybuzz I'm particularly unsure about this line. After we have found the conversation, we need to change the hallOfFame boolean from false to true and then update the database. My psuedo-code here would be more of the front-end, JavaScript way to accomplish this it but I'm sure that it's different when we're dealing with the back-end world. Would something along the lines of `user.userConversations[found].save(...)` work?

psybuzz replied:
I think your pseudo-code is the right way to start.  Although, it seems that the `found` variable is boolean, but if you have the `foundIndex`, you can indeed do `user.userConversations[foundIndex].hallOfFame == true;`, similar to as you suggested.

Once the value of the object is set to true, all that's left is to save the modified object in the database.  You could find an example of something like that in "routes/user.js" around line 38.  So, in the archive function, once we find the conversation in the user's list, you could do something like
```
user.save(function(err){
		if (err){
			return resError(res, "...some error message...");
		}

		/* Insert code that modified the isThrilled inside the convo object */
	});
```

Then, inside the save callback, if there is no error, we still need to update the convo object, which has a list of participants including whether or not each one is thrilled.  So in that place, you would need to: find the conversation using the conversationId you got before, update the participants property, and save the conversation.  The code flow could look like:
```
Conversation.findById(convoId, function(err, convo){
	if (err) return resError(res, "could not find convo")

	/* Do some regular JS that updates the convo.participants array to set the isThrilled to true on the correct user */

	// Save the modified conversation.
	convo.save(function(err){
		if (err) return resError(res, "could not save the convo")

		// Finally everything is done, so send the response back the the user with success.
		res.send({success: true, redirect: '/home'});
	});
});
```

Let me know if you have any questions at all!  I realize the code I wrote for some other functions isn't the must readable, so I can explain other functions if you need as well.  Thanks for taking this on, Nikil!

nikilselvam replied:
Great, thanks Erik! I'll take a look at this and let you know if I have any questions.

psybuzz commented:
Hey @nikilselvam, of course!  The client-side code you wrote looks good!  I'll post some further comments on lines in the server-side code.  You're on the right track :)

On promises, the main benefit is that they're supposed to reduce the depth of callbacks that we need to write, but in our case, I'm starting to feel like Q promises have made our codebase a lot harder to read, and might even be more confusing if we try and update code in the future.  In the spirit of keeping things maintainable, it might be good for us, in general, to start looking at other libraries that might better suit our asynchronous needs.

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +255,39 @@ exports.leave = function (req, res){
+	// id is found. 
+	User.findById( req.user._id, function (err, user){
+		if (err){
+			reject(res, "Could not find the user.");
```
I think the `reject(...)` here and on line 287 should actually be `resError(...)` calls.  Reject is used inside some promise functions, while resError is used for responding to the request with a error.

nikilselvam replied:
Ah, good call. I'll make the change!

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +255,39 @@ exports.leave = function (req, res){
+		if (found) {
+			// Something along the lines of: user.userConversations[found].hallOfFame == true;
+		} else{
+			reject(res, "Access denied.", "/error");
```
This `reject` should also be a `resError` I believe, since it responds to the user's request with an error.

nikilselvam commented:
@psybuzz I've made changes to update the correct participant's `isThrilled` attribute to true within the conversation object like you suggested above. Thanks for your help on that! Let me know what you think of the changes. 

One thing I noticed is that the current conversation entries in our database don't include `isThrilled` values for each person in the conversation's `participant` list. Could this mean that we currently don't set the `isThrilled` value to false when we add people to conversations? If so, should we add this to our logic so that our database is more explicit in whether its participants are thrilled or not about the conversation?

I also noticed one other potential issue that I'll address in the `Files Changed` section of this PR.

nikilselvam commented on routes/conversation.js:
```diff
@@ -169,6 +169,8 @@ exports.leave = function (req, res){
 			var userString = JSON.stringify(req.user._id);
 			var participantIdStrings = participantIds.map(function (id){return JSON.stringify(id)});
+			// Should this 'found' variable definition be changed to the equivalent 'foundIndex' variable
+			// definitions used below in 'exports.archive'?
```
@psybuzz I was looking through the leave functionality, and it seems like we use `found` as a boolean here and then delete one entry starting at index `found` to remove the target user from the conversation. If `found` is cast as a boolean, wouldn't the `convo.participants.splice(found, 1);` line inside `if (found){...}` always remove the first entry in the list? 

I wanted to bring this up because in the archive functionality below I'm saving `foundIndex` as an integer and then checking that this integer is valid (as opposed to storing a boolean). Should we adopt a similar approach here as well?

psybuzz replied:
Wow, that's definitely correct.  I think you've uncovered a bug!  The leave function should be using a foundIndex of sorts when splicing.  You're right, as it is now, it does seem like it would always remove the first entry, which is not what we want.  Let's change it!

nikilselvam replied:
Great! Should I change it in this PR or do you want to make the change in the 'leave functionality' PR that you're working on right now. Either way works for me!

psybuzz replied:
I can add that on to my 'leave functionality' PR, since it's a related bug.  Thanks for the discovery!

nikilselvam replied:
Sure thing! Thanks for making the change.

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +257,74 @@ exports.leave = function (req, res){
+					} else{
+						reject(res, "Access denied.", "/error");
+					}
+				});
```
After pulling from the branch and running it locally, I got an error around this line, so it may be that one of the brackets/parens isn't matching up.  I think there might need to be another `});` after this line here.

Hopefully we find a better way of coping with all these callbacks soon :p   There's a neat node module called "async", which looks promising, but it might be better to do a callback refactoring in the future, as I know that I've written a lot of callbacks in places that I shouldn't have!

nikilselvam replied:
Ah yes, good call. I'll add that in

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +257,74 @@ exports.leave = function (req, res){
+	// Search through the user's conversation list until the conversation with the correct
+	// id is found. 
+	User.findById( req.user._id, function (err, user){
+		if (err){
```
So, I found an error that's my fault, and I only realized it today.  When we do a database query and the results are empty, the query actually doesn't seem to throw an error.  Instead, it simply runs the callback with a null result.  For example, in this `User.findById` example, when a user is not found, err == null and user == null.

Therefore, everywhere we have `if (err)` and we really mean "if there was an error or the result is empty", we should really be writing `if (err || !user)`.  This is really my fault, I did not understand what the `err` variable really meant, I only assumed it included empty queries when it did not.  I'm sorry for making that assumption.  I'll be changing a lot of these instances in my PR, as well.  Could you please change this `if (err)` to `if (err || !user)`?

nikilselvam replied:
No worries, Erik! That seems like a reasonable assumption to make. I'll make the change here

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +257,74 @@ exports.leave = function (req, res){
+				// If there are no errors, proceed to update the isThrilled boolean of the user
+				// in the conversation object to true and save this object.
+				Conversation.findById(targetConversationId, function(err, convo){
+					if (err) return resError(res, "Could not find conversation");
```
I guess this `if (err)` should be a `if (err || !convo)` as well.  When the query finds no conversation, `err` is still null.

nikilselvam replied:
Ah yes, will do.

psybuzz commented on app.js:
```diff
@@ -93,6 +93,7 @@ app.post('/post/create', post.create);
 
 app.post('/user/delete', user.delete);
 app.post('/conversation/leave', conversation.leave);
+app.post('/conversaton/archive', conversation.archive);
```
Nit: "conversaton" --> "conversation"

nikilselvam replied:
Good catch. My bad on this one- that's a poor mistake to make

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +257,74 @@ exports.leave = function (req, res){
+
+				// If there are no errors, proceed to update the isThrilled boolean of the user
+				// in the conversation object to true and save this object.
+				Conversation.findById(targetConversationId, function(err, convo){
```
When doing a database query, it seems like it wants to use the original ID object as part of the query.  `targetConversationId` is no longer the object that Mongo expects because we used `JSON.stringify` to turn it into a string on line 278.  Thus, `targetConversationId` should probably be changed to the original id, `req.body.conversationId`.

I only know about this because I ran into an error during the summer that depended on it.  Took me forever to figure it out!

nikilselvam replied:
I see! That's a helpful tip- thanks! It does make sense that Mongo wouldn't accept a string rathe than the actual ID object

psybuzz commented:
To answer your question about setting isThrilled to false, you're right in that the values should be false, but they don't show up in the database.  I'm not sure why this is the case, because in routes/conversation.js, around line 70, there is a for loop that should be setting the isThrilled value to false for all the people.

I wonder if MongoDB might actually store all boolean-marked values as false by default.  If we put a `console.log(people[i].isThrilled)` inside that loop, it correctly prints `false` for all of them, so it seems that the behavior is correct, even though it doesn't show up in the DB.  As such, I think it's safe to just keep it as it is.

Overall, I had a couple minor comments,  but other than that, your code looks good!

nikilselvam commented:
@psybuzz I've addressed your comments in the latest commit. Thanks for your help! I'm learning a lot about back-end programming through this.

If the `isThrilled` values seem to be printing out correctly, then I agree that we're probably safe for now. Let's keep an eye on it when we start archiving conversations to ensure that the `isThrilled` boolean changes appropriately.

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +255,75 @@ exports.leave = function (req, res){
+							res.send({success: true, redirect: '/home'});
+						});
+					} else{
+						reject(res, "Access denied.", "/error");
```
Ahh, one more small thing, I think this `reject(...` should be a `return resError(...`.  Reject was used in the leave function because it used Promises, though since Promises aren't used here, we could handle errors with just the `resError`.

psybuzz commented on routes/conversation.js:
```diff
@@ -255,6 +255,75 @@ exports.leave = function (req, res){
+				});
+			});
+		} else{
+			reject(res, "Access denied.", "/error");
```
Same here, this should also be `return resError(...`

psybuzz commented:
Alright @nikilselvam !  I made a couple minor comments on lines using `reject(...`, but other than that, everything looks great!  After those are changed to `return resError`, this archiving backend functionality LGTM!

nikilselvam commented:
Thanks @psybuzz ! I've made those changes and then added two other `return resError` change earlier in the file in the last two commit. Would you mind taking one last look at the final two changes and make sure the extra `return resError`s look okay? 

@psybuzz @tonyxj Feel free to take one last look at this one. Once I get the okay from both of you I'll go ahead and merge this in!

tonyxj commented:
The code looks sublime @nikilselvam! It's amazing to see you guys go back and forth discussing various components. This is the foundation work for the Hall of Fame UI working coming up. I `isThrilled`. Merge it!

psybuzz commented:
Agreed, LGTM!

nikilselvam commented:
Great! Thanks guys. Ready to merge.
